File: app/src/main/java/app/fit/fitndflow/domain/common/usecase/CommonUseCase.java
package app.fit.fitndflow.domain.common.usecase;


import app.fit.fitndflow.domain.common.repository.CommonRepository;

public abstract class CommonUseCase<IN, OUT, REPO extends CommonRepository>{
    public REPO repository;

    public CommonUseCase(REPO commonRepo) {
        this.repository = commonRepo;
    }

    public abstract OUT executeUseCase(IN input);
}

File: app/src/main/java/app/fit/fitndflow/domain/common/usecase/CommonUseCase.kt
package app.fit.fitndflow.domain.common.usecase

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOn

abstract class UseCase<in Params : Any, out T>() {

    abstract fun run(params: Params): Flow<T>

    operator fun invoke(params: Params = Unit as Params) = run(params).flowOn(Dispatchers.IO)

}